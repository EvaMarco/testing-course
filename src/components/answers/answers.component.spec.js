import { AnswersComponent } from "./answers.component";
import { screen, waitFor } from "@testing-library/dom";
import { axe } from 'jest-axe';
import userEvent from '@testing-library/user-event'

const getBodyMain = () => {
  const main = document.createElement('main');
  document.body.appendChild(main);

  return main;
};

describe('Testing Answer component', () => {
  let answers;
  let correctAnswer;

  beforeAll(() => {
    answers = ['Madrid', 'Pekin', 'Bilbao', 'Kiev'];
    correctAnswer = 'Madrid';
    const main = getBodyMain();
    new AnswersComponent(main, answers, correctAnswer);
  });

  test('should match snapshot', () => {
    expect(document.body).toMatchSnapshot();
  });

  test('should pass a11y validation', async () => {
    expect(await axe(document.body)).toHaveNoViolations();
  });

  test('should render a button for each answer', () => {
		// Seleccionamos los elementos del body que tengan el rol "button"
    const buttons = screen.getAllByRole('button');
		// Comprobamos que son visibles y que contienen el texto que corresponde
    buttons.forEach((button, i) => {
      expect(button).toBeVisible();
      expect(button).toHaveTextContent(answers[i])
    });
  });

  describe('after click on a incorrect answer', () => {
    let answer;

    beforeAll(async () => {
      answer = screen.getByText('Bilbao');
      userEvent.click(answer);
    });

    test('should validate', () => {
      waitFor(() => expect(answer).toHaveClass('incorrect'));
    });

    test('should disable all buttons', () => {
      screen.getAllByRole('button').forEach((button) => {
        waitFor(() => expect(button).toBeDisabled());
      });
    });

    test('should show the correct answer', async () => {
      const correctButton = screen.getByText(correctAnswer)
      waitFor(() => expect(correctButton).toHaveClass('correct'));
    });
  });

  describe('after click on a correct answer', () => {
    let answer;

    beforeAll(() => {
      answer = screen.getByText(correctAnswer);
      userEvent.click(answer);
    });

    test('should validate', async () => {
      waitFor(() => expect(answer).toHaveClass('correct'));
    });
  });
});