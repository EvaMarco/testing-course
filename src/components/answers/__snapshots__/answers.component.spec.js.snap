// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Answer component should match snapshot 1`] = `
<body>
  <main>
    <button
      class="answer-button"
    >
      Madrid
    </button>
    <button
      class="answer-button"
    >
      Pekin
    </button>
    <button
      class="answer-button"
    >
      Bilbao
    </button>
    <button
      class="answer-button"
    >
      Kiev
    </button>
  </main>
</body>
`;
